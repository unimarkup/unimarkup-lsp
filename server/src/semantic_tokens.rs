use unimarkup_inline::Span;
use unimarkup_inline::NestedInline;
use unimarkup_inline::InlineKind;
use unimarkup_inline::Inline;
use unimarkup_core::elements::VerbatimBlock;
use unimarkup_core::elements::ParagraphBlock;
use unimarkup_core::elements::HeadingBlock;
use unimarkup_core::unimarkup::UnimarkupDocument;
use unimarkup_core::unimarkup_block::UnimarkupBlockKind;
use lsp_server::RequestId;
use lsp_types::SemanticToken;
use lsp_types::SemanticTokens;
use lsp_types::SemanticTokensResult;
use lsp_server::Response;
use lsp_types::SemanticTokensParams;

pub fn get_semantic_tokens(id: RequestId, params: SemanticTokensParams, rendered_um: Option<&UnimarkupDocument>) -> Response {
	eprintln!("got semantic token request #{}: {:?}", id, params);
	let mut tokens = SemanticTokens {
		result_id: Some(id.to_string()),
		..Default::default()
	};

	if let Some(um_doc) = rendered_um {
		tokens.data = make_relative(um_doc.tokens());
		eprintln!("Sent tokens: {:?}", tokens.data);
	}

	let result = Some(SemanticTokensResult::Tokens(tokens));

	let result = serde_json::to_value(&result).unwrap();
	Response { id, result: Some(result), error: None }
}

/// Brings all tokens in relative position offsets
fn make_relative(mut tokens: Vec<SemanticToken>) -> Vec<SemanticToken> {
	tokens.sort_by(|b, a| {
		let outer = a.delta_line.partial_cmp(&b.delta_line);
		if let Some(outer_order) = outer {
			if outer_order == std::cmp::Ordering::Equal {
				return a.delta_start.partial_cmp(&b.delta_start).unwrap();
			} else {
				return outer_order;
			}
		}
		std::cmp::Ordering::Equal
	});

	let mut sorted_tokens = tokens.clone();
	eprintln!("Tokens: {:?}", sorted_tokens);
	for (i, token) in sorted_tokens.iter_mut().enumerate() {
		if i < tokens.len() - 1 {
			if let Some(next_token) = tokens.get(i + 1) {
				token.delta_line = token.delta_line - next_token.delta_line;
				if token.delta_line == 0 {
					token.delta_start = token.delta_start - next_token.delta_start;
				}
			}
		}
	}

	sorted_tokens.reverse();
	sorted_tokens
}

trait SemanticBlockTokenizer {
	fn tokens(&self) -> Vec<SemanticToken>;
}

impl SemanticBlockTokenizer for UnimarkupDocument {
	fn tokens(&self) -> Vec<SemanticToken> { 
		let mut tokens = Vec::<SemanticToken>::new();
		for block in &self.elements {
			tokens.append(&mut block.tokens());
		}
		tokens
	}
}

impl SemanticBlockTokenizer for UnimarkupBlockKind {
	fn tokens(&self) -> Vec<SemanticToken> { 
		match self {
			UnimarkupBlockKind::Heading(heading) => {
				return heading.tokens();
			},
			UnimarkupBlockKind::Paragraph(paragraph) => {
				return paragraph.tokens();
			},
			UnimarkupBlockKind::Verbatim(verbatim) => {
				return verbatim.tokens();
			}
		}	
	}
}

impl SemanticBlockTokenizer for HeadingBlock {
	fn tokens(&self) -> Vec<SemanticToken> {
		let mut open_tokens = vec![(3,0)];
		let mut tokens = self.content.tokens(&mut open_tokens);
		let last_inline = self.content.last().unwrap();
		let last_pos;
		match last_inline {
			InlineKind::Bold(nested)
			| InlineKind::Italic(nested)
			| InlineKind::BoldItalic(nested) => { last_pos = nested.span.end },
			InlineKind::Verbatim(flat)
			| InlineKind::Plain(flat)
			| InlineKind::PlainNewLine(flat)
			| InlineKind::EscapedNewLine(flat)
			| InlineKind::EscapedSpace(flat) => { last_pos = flat.span.end },
		};

		tokens.push(
			SemanticToken{
				delta_line: to_lsp_line_nr(last_pos.line),
				delta_start: 0,
				length: last_pos.column as u32,
				token_type: 3,
				token_modifiers_bitset: 0
			}
		);

		tokens
	}
}


impl SemanticBlockTokenizer for ParagraphBlock {
	fn tokens(&self) -> Vec<SemanticToken> {
		let mut open_tokens = vec![].into();
		self.content.tokens(&mut open_tokens)
	}
}

impl SemanticBlockTokenizer for VerbatimBlock {
	fn tokens(&self) -> Vec<SemanticToken> {
		let mut tokens = vec![
			SemanticToken{
				delta_line: to_lsp_line_nr(self.line_nr),
				delta_start: 0,
				length: 10,
				token_type: self.token_type(),
				token_modifiers_bitset: 0
			}
		];

		let lines = self.content.lines();
		for (i, line) in lines.enumerate() {
			tokens.push(
				SemanticToken{
					delta_line: to_lsp_line_nr(self.line_nr + i + 1),
					delta_start: 0,
					length: (line.len() as u32),
					token_type: self.token_type(),
					token_modifiers_bitset: 0
				}
			);
		}

		tokens.push(
			SemanticToken{
				delta_line: to_lsp_line_nr(self.line_nr + self.content.lines().count() + 1),
				delta_start: 0,
				length: 10,
				token_type: self.token_type(),
				token_modifiers_bitset: 0
			}
		);

		tokens
	}
}

/// Function converts a Unimarkup line number to LSP by starting at 0
fn to_lsp_line_nr(line_nr: usize) -> u32 {
	let nr = line_nr as u32;
	nr - 1
}

trait SemanticInlineTokenizer {
	fn tokens(&self, open_token_types: &mut Vec<(u32, u32)>) -> Vec<SemanticToken>;
}

impl SemanticInlineTokenizer for Inline {
	fn tokens(&self, open_token_types: &mut Vec<(u32, u32)>) -> Vec<SemanticToken> {
		let mut tokens = Vec::<SemanticToken>::new();
		for inlines in self {
			tokens.append(&mut inlines.tokens(open_token_types))
		}

		tokens
	}
}

impl SemanticInlineTokenizer for InlineKind {
	fn tokens(&self, open_token_types: &mut Vec<(u32, u32)>) -> Vec<SemanticToken> {
		
		match self {
			InlineKind::Bold(nested)
			| InlineKind::Italic(nested) => {
				open_token_types.push((self.token_type(), nested.span.start.column as u32));

				let mut nested_tokens = nested.tokens(open_token_types);
				nested_tokens.push(
					SemanticToken{ 
						delta_line: to_lsp_line_nr(nested.span.end.line),
						delta_start: get_delta_start(nested.span),
						length: get_token_length(nested.span),
						token_type: self.token_type(),
						..Default::default()
					}
				);

				open_token_types.pop();

				return nested_tokens;
			},
			InlineKind::Plain(_) => [].into(),
			InlineKind::PlainNewLine(flat) => {
				let mut tokens = Vec::new();
				for (token_type, start_col) in open_token_types.iter_mut() {
					tokens.push(
						SemanticToken{
							delta_line: to_lsp_line_nr(flat.span.start.line),
							delta_start: *start_col,
							length: flat.span.start.column as u32,
							token_type: *token_type,
							..Default::default()
						}
					);

					if *start_col != 0 {
						*start_col = 0;
					}
				}

				tokens
			},
			_ => {
				[].into()
			}
		}
	}
}

fn get_delta_start(span: Span) -> u32 {
	if span.start.line == span.end.line {
		span.start.column as u32
	} else {
		0
	}
}

fn get_token_length(span: Span) -> u32 {
	if span.start.line == span.end.line {
		(span.end.column - span.start.column) as u32
	} else {
		span.end.column as u32
	}
}

impl SemanticInlineTokenizer for NestedInline {
	fn tokens(&self, open_token_types: &mut Vec<(u32, u32)>) -> Vec<SemanticToken> {
		let mut tokens = Vec::<SemanticToken>::new();
		for nested in &self.content {
			tokens.append(&mut nested.tokens(open_token_types))
		}
		tokens
	}
}



trait SemanticTokenMapType {
	fn token_type(&self) -> u32;
}

const NO_TOKEN_TYPE: u32 = u32::max_value();

impl SemanticTokenMapType for InlineKind {	
	fn token_type(&self) -> u32 {
		match self {
			InlineKind::Bold(_) => 1,
			InlineKind::Italic(_) => 2,
			InlineKind::Verbatim(_) => 18,
			_ => NO_TOKEN_TYPE,
		}
	}
}

impl SemanticTokenMapType for UnimarkupBlockKind {
	fn token_type(&self) -> u32 {
		match self {
			UnimarkupBlockKind::Heading(_) => 16,
			UnimarkupBlockKind::Verbatim(_) => 18,
			_ => NO_TOKEN_TYPE,
		}
	}
}

impl SemanticTokenMapType for VerbatimBlock {
	fn token_type(&self) -> u32 {
		18
	}
}
