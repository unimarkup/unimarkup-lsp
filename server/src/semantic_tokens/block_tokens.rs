use lsp_types::SemanticToken;
use unimarkup_core::{
    elements::{HeadingBlock, ParagraphBlock, VerbatimBlock},
    unimarkup::UnimarkupDocument,
    unimarkup_block::UnimarkupBlockKind,
};

use super::{
    delta_conversions::to_lsp_line_nr, inline_tokens::SemanticInlineTokenizer, OpenTokenType,
    TokenType, TokenValue,
};

pub(crate) trait SemanticBlockTokenizer {
    fn tokens(&self, open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken>;
}

impl SemanticBlockTokenizer for UnimarkupDocument {
    fn tokens(&self, open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken> {
        let mut tokens = Vec::<SemanticToken>::new();
        for block in &self.elements {
            tokens.append(&mut block.tokens(open_types));
        }
        tokens
    }
}

impl SemanticBlockTokenizer for UnimarkupBlockKind {
    fn tokens(&self, open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken> {
        match self {
            UnimarkupBlockKind::Heading(heading) => heading.tokens(open_types),
            UnimarkupBlockKind::Paragraph(paragraph) => paragraph.tokens(open_types),
            UnimarkupBlockKind::Verbatim(verbatim) => verbatim.tokens(open_types),
        }
    }
}

impl SemanticBlockTokenizer for HeadingBlock {
    fn tokens(&self, open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken> {
        let mut tokens = vec![SemanticToken {
            delta_line: to_lsp_line_nr(self.line_nr),
            delta_start: calculate_column_offset(open_types),
            length: (u8::from(self.level) + 1).into(), // +1 for space
            token_type: TokenType::Heading.value(),
            token_modifiers_bitset: 0,
        }];
        tokens.append(&mut self.content.tokens(&TokenType::Heading, &mut vec![]));
        tokens
    }
}

impl SemanticBlockTokenizer for ParagraphBlock {
    fn tokens(&self, _open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken> {
        self.content.tokens(&TokenType::Paragraph, &mut vec![])
    }
}

impl SemanticBlockTokenizer for VerbatimBlock {
    fn tokens(&self, _open_types: &mut Vec<OpenTokenType>) -> Vec<SemanticToken> {
        //TODO: Change length after VerbatimBlock contains needed information
        let mut tokens = vec![SemanticToken {
            delta_line: to_lsp_line_nr(self.line_nr),
            delta_start: 0,
            length: 50,
            token_type: TokenType::VerbatimBlock.value(),
            token_modifiers_bitset: 0,
        }];

        let lines = self.content.lines();
        for (i, line) in lines.enumerate() {
            tokens.push(SemanticToken {
                delta_line: to_lsp_line_nr(self.line_nr + i + 1),
                delta_start: 0,
                length: (line.len() as u32),
                token_type: TokenType::VerbatimBlock.value(),
                token_modifiers_bitset: 0,
            });
        }

        tokens.push(SemanticToken {
            delta_line: to_lsp_line_nr(self.line_nr + self.content.lines().count() + 1),
            delta_start: 0,
            length: 50,
            token_type: TokenType::VerbatimBlock.value(),
            token_modifiers_bitset: 0,
        });

        tokens
    }
}

fn calculate_column_offset(open_types: &mut [OpenTokenType]) -> u32 {
    match open_types.last() {
        Some(last_open) => {
            if let Some(length) = last_open.length {
                length + last_open.start_column_offset
            } else {
                last_open.start_column_offset
            }
        }
        None => 0,
    }
}
