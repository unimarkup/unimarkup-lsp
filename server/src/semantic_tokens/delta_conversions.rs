
/// Function converts a Unimarkup line number to LSP by starting at 0
pub(crate) fn to_lsp_line_nr(line_nr: usize) -> u32 {
	let nr = line_nr as u32;
	nr - 1
}
