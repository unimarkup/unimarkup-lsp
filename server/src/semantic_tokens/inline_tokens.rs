use lsp_types::SemanticToken;
use unimarkup_inline::{Inline, NestedContent, TokenDelimiters, TokenKind};

use super::{
    delta_conversions::to_lsp_line_nr, OpenTokenModifier, TokenModifier, TokenType, TokenValue,
    NO_TOKEN_TYPE,
};

pub(crate) trait SemanticInlineTokenizer {
    fn tokens(
        &self,
        token_type: &TokenType,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken>;
}

impl SemanticInlineTokenizer for NestedContent {
    fn tokens(
        &self,
        token_type: &TokenType,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken> {
        self.iter()
            .flat_map(|inline| inline.tokens(token_type, open_modifiers))
            .collect()
    }
}

impl SemanticInlineTokenizer for Inline {
    fn tokens(
        &self,
        token_type: &TokenType,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken> {
        match self {
            Inline::Bold(nested) | Inline::Italic(nested) => {
                open_modifiers.push(self.into());
                let delimiters = self.delimiters();

                let mut tokens = vec![SemanticToken {
                    delta_line: to_lsp_line_nr(self.span().start().line),
                    delta_start: self.span().start().column as u32,
                    length: delimiters.open().as_str().len() as u32,
                    token_type: token_type.value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                }];

                tokens.append(&mut nested.tokens(token_type, open_modifiers));

                tokens.push(SemanticToken {
                    delta_line: to_lsp_line_nr(self.span().end().line),
                    delta_start: (self.span().end().column
                        - delimiters
                            .close()
                            .unwrap_or(TokenKind::Plain)
                            .as_str()
                            .len()) as u32,
                    length: delimiters
                        .close()
                        .unwrap_or(TokenKind::Plain)
                        .as_str()
                        .len() as u32,
                    token_type: token_type.value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                });

                open_modifiers.pop();
                tokens
            }
            Inline::Plain(plain_content) => {
                if token_type.value() != NO_TOKEN_TYPE || !open_modifiers.is_empty() {
                    vec![SemanticToken {
                        delta_line: to_lsp_line_nr(plain_content.span().start().line),
                        delta_start: plain_content.span().start().column as u32,
                        length: (plain_content.span().end().column
                            - plain_content.span().start().column)
                            as u32,
                        token_type: token_type.value(),
                        token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                    }]
                } else {
                    vec![]
                }
            }
            Inline::TextGroup(nested) => {
                let delimiters: TokenDelimiters = self.delimiters();

                let mut tokens = vec![SemanticToken {
                    delta_line: to_lsp_line_nr(nested.span().start().line),
                    delta_start: nested.span().start().column as u32,
                    length: delimiters.open().as_str().len() as u32,
                    token_type: token_type.value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                }];

                tokens.append(&mut nested.tokens(token_type, open_modifiers));

                tokens.push(SemanticToken {
                    delta_line: to_lsp_line_nr(nested.span().end().line),
                    delta_start: (nested.span().end().column
                        - delimiters
                            .close()
                            .unwrap_or(TokenKind::Plain)
                            .as_str()
                            .len()) as u32,
                    length: delimiters
                        .close()
                        .unwrap_or(TokenKind::Plain)
                        .as_str()
                        .len() as u32,
                    token_type: token_type.value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                });

                tokens
            }
            _ => vec![],
        }
    }
}

// impl TokenValue for InlineKind {
//     fn value(&self) -> u32 {
//         match self {
//             InlineKind::Bold(_) => TokenModifier::Bold.value(),
//             InlineKind::Italic(_) => TokenModifier::Italic.value(),
//             InlineKind::BoldItalic(_) => TokenModifier::BoldItalic.value(),
//             InlineKind::Verbatim(_) => TokenModifier::Verbatim.value(),
//             _ => NO_TOKEN_TYPE,
//         }
//     }
// }

impl From<&Inline> for OpenTokenModifier {
    fn from(inline: &Inline) -> Self {
        match inline {
            Inline::Bold(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Bold,
                ..Default::default()
            },
            Inline::Italic(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Italic,
                ..Default::default()
            },
            Inline::Verbatim(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Verbatim,
                ..Default::default()
            },
            _ => OpenTokenModifier {
                ..Default::default()
            },
        }
    }
}

fn get_modifier_bitfield(modifiers: &Vec<OpenTokenModifier>) -> u32 {
    let mut field = 0;

    for modifier in modifiers {
        field |= modifier.token_modifier.value();
    }

    field
}
