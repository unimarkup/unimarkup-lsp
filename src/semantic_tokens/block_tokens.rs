use lsp_types::SemanticToken;
use unimarkup_core::{
    document::Document,
    elements::blocks::Block,
    elements::{atomic::Heading, atomic::Paragraph, enclosed::Verbatim},
};

use super::{
    inline_tokens::SemanticInlineTokenizer, TokenValue,
};

// #[derive(Debug, Default, Clone)]
// pub(crate) struct OpenTokenType {
//     /// The open token type
//     _token_type: TokenType,
//     /// Column offset the content of this type starts.
//     /// Needed for nested blocks.
//     ///
//     /// e.g. verbatim inside list
//     start_column_offset: u32,
//     /// The type length, or `None` if the type goes to end of line
//     length: Option<u32>,
// }

#[derive(Default, Debug, Clone)]
pub(crate) enum TokenType {
    #[default]
    Paragraph,
    Heading,
    Verbatim,
}

impl TokenValue for TokenType {
    fn value(&self) -> u32 {
        match self {
            TokenType::Paragraph => 2,//21,
            TokenType::Heading => 3,
            TokenType::Verbatim => 18,
        }
    }
}

pub(crate) trait SemanticBlockTokenizer {
    fn tokens(&self) -> Vec<SemanticToken>;
}

impl SemanticBlockTokenizer for Document {
    fn tokens(&self) -> Vec<SemanticToken> {
        let mut tokens = Vec::<SemanticToken>::new();
        for block in &self.blocks {
            tokens.append(&mut block.tokens());
        }
        tokens
    }
}

impl SemanticBlockTokenizer for Block {
    fn tokens(&self) -> Vec<SemanticToken> {
        match self {
            Block::Heading(heading) => heading.tokens(),
            Block::Paragraph(paragraph) => paragraph.tokens(),
            Block::Verbatim(verbatim) => verbatim.tokens(),
            _ => todo!(),
        }
    }
}

impl SemanticBlockTokenizer for Heading {
    fn tokens(&self) -> Vec<SemanticToken> {
        let mut tokens = vec![SemanticToken {
            delta_line: self.line_nr as u32,
            delta_start: 1,
            length: (u8::from(self.level)).into(),
            token_type: TokenType::Heading.value(),
            token_modifiers_bitset: 0,
        }];

        tokens.append(
            &mut self
                .content
                .iter()
                .flat_map(|inline| inline.tokens(&mut vec![]))
                .collect(),
        );

        dbg!(tokens)
    }
}

impl SemanticBlockTokenizer for Paragraph {
    fn tokens(&self) -> Vec<SemanticToken> {
        self.content
            .iter()
            .flat_map(|inline| inline.tokens(&mut vec![]))
            .collect()
    }
}

impl SemanticBlockTokenizer for Verbatim {
    fn tokens(&self) -> Vec<SemanticToken> {
        //TODO: Change length after Verbatim contains needed information
        let mut tokens = vec![SemanticToken {
            delta_line: self.line_nr as u32,
            delta_start: 1,
            length: 50,
            token_type: TokenType::Verbatim.value(),
            token_modifiers_bitset: 0,
        }];

        let lines = self.content.lines();
        for (i, line) in lines.enumerate() {
            tokens.push(SemanticToken {
                delta_line: (self.line_nr + i + 1) as u32,
                delta_start: 1,
                length: (line.len() as u32),
                token_type: TokenType::Verbatim.value(),
                token_modifiers_bitset: 0,
            });
        }

        tokens.push(SemanticToken {
            delta_line: (self.line_nr + self.content.lines().count() + 1) as u32,
            delta_start: 1,
            length: 50,
            token_type: TokenType::Verbatim.value(),
            token_modifiers_bitset: 0,
        });

        tokens
    }
}

// fn calculate_column_offset(open_types: &mut [OpenTokenType]) -> u32 {
//     match open_types.last() {
//         Some(last_open) => {
//             if let Some(length) = last_open.length {
//                 length + last_open.start_column_offset
//             } else {
//                 last_open.start_column_offset
//             }
//         }
//         None => 0,
//     }
// }
