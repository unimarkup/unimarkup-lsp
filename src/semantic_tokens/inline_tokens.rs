use lsp_types::SemanticToken;
use unimarkup_inline::{Inline, NestedContent, TokenDelimiters, TokenKind};

use super::{
    block_tokens::TokenType, delta_conversions::to_lsp_line_nr, TokenValue,
};

#[derive(Debug, Default, Clone)]
pub(crate) struct OpenTokenModifier {
    /// The open token modifier
    token_modifier: TokenModifier,
    /// Column start of the modifier
    _start_column: u32,
    /// Column offset the content of this modifier starts.
    /// Needed for nested blocks.
    ///
    /// e.g. verbatim inside list
    _start_column_offset: u32,
}

#[derive(Default, Debug, Clone)]
pub(crate) enum TokenModifier {
    #[default]
    Plain,
    Bold,
    Italic,
    Verbatim,
}

impl TokenValue for TokenModifier {
    fn value(&self) -> u32 {
        // Note: These values must set the correct modifier bit
        match self {
            TokenModifier::Plain => 0,
            TokenModifier::Bold => 1,
            TokenModifier::Italic => 1 << 1,
            TokenModifier::Verbatim => 3,
        }
    }
}

pub(crate) trait SemanticInlineTokenizer {
    fn tokens(
        &self,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken>;
}

impl SemanticInlineTokenizer for NestedContent {
    fn tokens(
        &self,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken> {
        self.iter()
            .flat_map(|inline| inline.tokens(open_modifiers))
            .collect()
    }
}

impl SemanticInlineTokenizer for Inline {
    fn tokens(
        &self,
        open_modifiers: &mut Vec<OpenTokenModifier>,
    ) -> Vec<SemanticToken> {
        match self {
            Inline::Bold(nested) | Inline::Italic(nested) => {
                open_modifiers.push(self.into());
                let delimiters = self.delimiters();

                let mut tokens = vec![SemanticToken {
                    delta_line: self.span().start().line as u32,
                    delta_start: self.span().start().column as u32,
                    length: delimiters.open().as_str().len() as u32,
                    token_type: TokenType::default().value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                }];

                tokens.append(&mut nested.tokens(open_modifiers));

                let closing_delim = delimiters
                    .close()
                    .expect("Could not unwrap non-existent closing tag");

                tokens.push(SemanticToken {
                    delta_line: self.span().end().line as u32,
                    delta_start: (self.span().end().column + 1 - closing_delim.as_str().len()) as u32,
                    length: closing_delim.as_str().len() as u32,
                    token_type: TokenType::default().value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                });

                open_modifiers.pop();
                tokens
            }
            Inline::Verbatim(plain_content) => {
                open_modifiers.push(self.into());

                let tokens = vec![SemanticToken {
                    delta_line: plain_content.span().start().line as u32,
                    delta_start: plain_content.span().start().column as u32,
                    length: plain_content.content_len() as u32,
                    token_type: TokenType::default().value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                }];

                open_modifiers.pop();

                tokens
            }
            Inline::Plain(plain_content) => {
                dbg!(plain_content.span().start().line);
                if !open_modifiers.is_empty() {
                    vec![SemanticToken {
                        delta_line: plain_content.span().start().line as u32,
                        delta_start: plain_content.span().start().column as u32,
                        length: plain_content.content_len() as u32,
                        token_type: TokenType::default().value(),
                        token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                    }]
                } else {
                    vec![]
                }
            }
            Inline::TextGroup(nested) => {
                let delimiters: TokenDelimiters = self.delimiters();

                let mut tokens = vec![SemanticToken {
                    delta_line: nested.span().start().line as u32,
                    delta_start: nested.span().start().column as u32,
                    length: delimiters.open().as_str().len() as u32,
                    token_type: TokenType::default().value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                }];

                tokens.append(&mut nested.tokens(open_modifiers));

                tokens.push(SemanticToken {
                    delta_line: to_lsp_line_nr(nested.span().end().line),
                    delta_start: (nested.span().end().column
                        - delimiters
                            .close()
                            .unwrap_or(TokenKind::Plain)
                            .as_str()
                            .len()) as u32,
                    length: delimiters
                        .close()
                        .unwrap_or(TokenKind::Plain)
                        .as_str()
                        .len() as u32,
                    token_type: TokenType::default().value(),
                    token_modifiers_bitset: get_modifier_bitfield(open_modifiers),
                });

                tokens
            }
            _ => vec![],
        }
    }
}

impl TokenValue for Inline {
    fn value(&self) -> u32 {
        match self {
            Inline::Bold(_) => TokenModifier::Bold.value(),
            Inline::Italic(_) => TokenModifier::Italic.value(),
            Inline::Verbatim(_) => TokenModifier::Verbatim.value(),
            _ => TokenModifier::default().value(),
        }
    }
}

// impl From<&Inline> for TokenType {
//     fn from(inline: &Inline) -> Self {
//         match *inline {
//             Inline::Bold(_) => TokenType::Bold,
//             Inline::Italic(_) => TokenType::Italic,
//             Inline::Verbatim(_) => TokenType::Verbatim,
//             _ => TokenType::Paragraph,
//         }
//     }
// }

impl From<&Inline> for OpenTokenModifier {
    fn from(inline: &Inline) -> Self {
        match inline {
            Inline::Bold(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Bold,
                ..Default::default()
            },
            Inline::Italic(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Italic,
                ..Default::default()
            },
            Inline::Verbatim(_) => OpenTokenModifier {
                token_modifier: TokenModifier::Verbatim,
                ..Default::default()
            },
            _ => OpenTokenModifier {
                ..Default::default()
            },
        }
    }
}

fn get_modifier_bitfield(modifiers: &Vec<OpenTokenModifier>) -> u32 {
    let mut field = 0;

    for modifier in modifiers {
        field |= modifier.token_modifier.value();
    }

    field
}
