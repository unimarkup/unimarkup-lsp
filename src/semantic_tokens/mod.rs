use lsp_server::RequestId;
use lsp_server::Response;
use lsp_types::SemanticToken;
use lsp_types::SemanticTokens;
use lsp_types::SemanticTokensParams;
use lsp_types::SemanticTokensResult;
use unimarkup_core::document::Document;

use self::block_tokens::SemanticBlockTokenizer;

mod block_tokens;
mod delta_conversions;
mod inline_tokens;

trait TokenValue {
    fn value(&self) -> u32;
}

pub fn get_semantic_tokens_response(
    id: RequestId,
    _params: SemanticTokensParams,
    document: Option<&Document>,
) -> Response {
    let mut tokens = SemanticTokens {
        result_id: Some(id.to_string()),
        ..Default::default()
    };

    if let Some(um_doc) = document {
        tokens.data = get_semantic_tokens(um_doc);
    }

    let result = Some(SemanticTokensResult::Tokens(tokens));
    let result = serde_json::to_value(&result).unwrap();
    Response {
        id,
        result: Some(result),
        error: None,
    }
}

pub fn get_semantic_tokens(document: &Document) -> Vec<SemanticToken> {
    make_relative(document.tokens())
}

/// Brings all tokens in relative position offsets
fn make_relative(mut tokens: Vec<SemanticToken>) -> Vec<SemanticToken> {
    // Bring tokens in descending order (highest line/column at first)
    tokens.sort_by(|a, b| {
        let outer = b.delta_line.partial_cmp(&a.delta_line);
        if let Some(outer_order) = outer {
            if outer_order == std::cmp::Ordering::Equal {
                return b.delta_start.partial_cmp(&a.delta_start).unwrap();
            } else {
                return outer_order;
            }
        }
        std::cmp::Ordering::Equal
    });

    // Note: unimarkup positions start with line = 1 and column = 1
    // LSP positions start with line = 0 and column = 0
    // => `-1` to correct this
    let mut sorted_tokens = tokens.clone();
    for (i, token) in sorted_tokens.iter_mut().enumerate() {
        if i < tokens.len() - 1 {
            if let Some(next_token) = tokens.get(i + 1) {
                token.delta_line -= next_token.delta_line;
                if token.delta_line == 0 {
                    token.delta_start -= next_token.delta_start;
                } else {
                    token.delta_start -= 1;
                }              
            }
        } else {
            token.delta_start -= 1;
            token.delta_line -= 1;
        }  
    }

    sorted_tokens.reverse();
    sorted_tokens
}
