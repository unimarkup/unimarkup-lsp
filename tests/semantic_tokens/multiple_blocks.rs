use insta::assert_snapshot;
use unimarkup_core::commons::config::Config;
use unimarkup_lsp::semantic_tokens::get_semantic_tokens;

#[test]
fn bold_token_modifier_in_two_blocks() {
    let input = "**bold text**\n\n**other bold text**";
    let um = unimarkup_core::Unimarkup::parse(input, Config::default());

    let tokens = get_semantic_tokens(&um);

    assert_snapshot!(format!("{:#?}", tokens));
}
