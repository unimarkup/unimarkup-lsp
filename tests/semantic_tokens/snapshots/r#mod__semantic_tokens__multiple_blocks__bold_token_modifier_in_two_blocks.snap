---
source: tests/semantic_tokens/multiple_blocks.rs
expression: "format!(\"{:#?}\", tokens)"
---
[
    SemanticToken {
        delta_line: 0,
        delta_start: 0,
        length: 2,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
    SemanticToken {
        delta_line: 0,
        delta_start: 2,
        length: 9,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
    SemanticToken {
        delta_line: 0,
        delta_start: 9,
        length: 2,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
    SemanticToken {
        delta_line: 2,
        delta_start: 0,
        length: 2,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
    SemanticToken {
        delta_line: 0,
        delta_start: 2,
        length: 15,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
    SemanticToken {
        delta_line: 0,
        delta_start: 15,
        length: 2,
        token_type: 2,
        token_modifiers_bitset: 1,
    },
]
